mm2src/coins/eth/wallet_connect.rs

```diff
@@ -0,0 +1,267 @@
+            "Waited too long until the transaction {tx_hash:?} appear on the RPC node"
+        ))))?;
+        let tx_hex = BytesJson::from(rlp::encode(&signed_tx).to_vec());
+
+        Ok((signed_tx, tx_hex))
+    }
+
+    fn session_topic(&self) -> Result<&str, Self::Error> {
+        if let EthPrivKeyPolicy::WalletConnect { ref session_topic, .. } = &self.priv_key_policy {
+            return Ok(session_topic);
+        }
+
+        MmError::err(EthWalletConnectError::SessionError(format!(
+            "{} is not activated via WalletConnect",
+            self.ticker()
+        )))
+    }
+}
+
+pub async fn eth_request_wc_personal_sign(
+    wc: &WalletConnectCtx,
+    session_topic: &str,
+    chain_id: u64,
+) -> MmResult<(H520, Address), EthWalletConnectError> {
+    let chain_id = WcChainId::new_eip155(chain_id.to_string());
+    wc.validate_update_active_chain_id(session_topic, &chain_id).await?;
+
+    let (account_str, _) = wc.get_account_and_properties_for_chain_id(session_topic, &chain_id)?;
+    let message = "Authenticate with Komodefi";
+    let params = {
+        let message_hex = format!("0x{}", hex::encode(message));
+        json!(&[&message_hex, &account_str])
+    };
+    let data = wc
+        .send_session_request_and_wait::<String>(session_topic, &chain_id, WcRequestMethods::PersonalSign, params)
+        .await?;
+
+    Ok(extract_pubkey_from_signature(&data, message, &account_str)
+        .mm_err(|err| WalletConnectError::SessionError(err.to_string()))?)
+}
+
+fn extract_pubkey_from_signature(
+    signature_str: &str,
+    message: &str,
+    account: &str,
+) -> MmResult<(H520, Address), EthWalletConnectError> {
+    let account =
+        H160::from_str(&account[2..]).map_to_mm(|err| EthWalletConnectError::InternalError(err.to_string()))?;
+    let uncompressed: H520 = {
+        let message_hash = hash_message(message);
+        let signature = Signature::from_str(&signature_str[2..])
+            .map_to_mm(|err| EthWalletConnectError::InvalidSignature(err.to_string()))?;
+        let pubkey = recover(&signature, &message_hash).map_to_mm(|err| {
+            EthWalletConnectError::InvalidSignature(format!(
+                "Couldn't recover public key from the signature: '{signature:?}, error: {err:?}'"
+            ))
+        })?;
+        pubkey.serialize_uncompressed().into()
+    };
+
+    let mut public = Public::default();
+    public.as_mut().copy_from_slice(&uncompressed[1..65]);
+
+    let recovered_address = public_to_address(&public);
+    if account != recovered_address {
+        return MmError::err(EthWalletConnectError::AccoountMisMatch(format!(
+            "Recovered address '{recovered_address:?}' should be the same as '{account:?}'"
+        )));
+    }
+
+    Ok((uncompressed, recovered_address))
+}
+
+pub(crate) fn recover(signature: &Signature, message: &Message) -> Result<PublicKey, ethkey::Error> {
+    let recovery_id = {
+        let recovery_id = (signature[64] as i32)
```

Comment:
You can use `signature.get(64)` for safety and convert `None` to `ethkey::Error::InvalidSignature`.

Reply:
hmmm, this adds unnecessary bound checking that could be avoided since `Signature` itself can only have a length `64`

definition of `Signature`
```rust
pub struct Signature([u8; 65]);
```

Reply:
Ah, didn't know the details of `Signature`, thanks.